NSFS needs to handle ENOTDIR

### Environment info
- NooBaa Version: **5.15.2**
- Platform: **NA**

### Actual behavior
1. Upload an object to a bucket - `s3 cp file s3://bucket/file`
2. Using the key `file/` - note the ending slash...
3. S3 requests in namespace_fs fails with InternalError because ENOTDIR is not handled (by all ops besides delete-object).
4. This was observed when running `spark.read.parquet('s3a://bucket/file')` (in pyspark.sql) which uses list-objects to check if the path refers to single object or a list of objects (dir) but does not expect to get back an internal error in any case.

### Expected behavior
1. Should not fail with InternalError
2. Instead this should be generally treated as ENOENT is treated by the different ops (list would just return empty, head/get should return 404, put should allow to put).
3. There are multiple code flows that throw and needs fixing, just to name a few common ones - check_bucket_boundaries, check_access, process_dir calling dir_cache calling stat, etc...

### Steps to reproduce
```
s3 cp file s3://bucket/file
s3api list-objects --bucket bucket --prefix file/ --delimiter /
s3api head-object --bucket bucket --key file/
s3api get-object --bucket bucket --key file/ outfile
s3api put-object --bucket bucket --key file/
...
```

### More information - Screenshots / Logs / Other output

This is the error from head-object and put-object after handling check_bucket_boundaries and check_access with a local fix -

`Apr 15 19:37:20 nsfs node[249503]: Apr-15 19:37:20.373 [nsfs/249503] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Resource>/guy/policy.json/</Resource><RequestId>lv1lh88j-bog5lr-15de</RequestId></Error> HEAD /guy/policy.json/ {"host":"localhost","accept-encoding":"identity","user-agent":"aws-cli/2.13.32 Python/3.11.6 Linux/4.18.0-372.57.1.el8_6.x86_64 exe/x86_64.rhel.8 prompt/off command/s3api.head-object","x-amz-date":"20240415T233720Z","x-amz-content-sha256":"UNSIGNED-PAYLOAD","authorization":"AWS4-HMAC-SHA256 ..."} Error: Not a directory`

`Apr 15 19:44:32 nsfs node[249503]: Apr-15 19:44:32.973 [nsfs/249503] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Resource>/guy/policy.json/</Resource><RequestId>lv1lqi11-4ynxkx-10b2</RequestId></Error> PUT /guy/policy.json/ {"host":"localhost","accept-encoding":"identity","user-agent":"aws-cli/2.13.32 Python/3.11.6 Linux/4.18.0-372.57.1.el8_6.x86_64 exe/x86_64.rhel.8 prompt/off command/s3api.put-object","content-md5":"Tle6aiCWLjF/nSAursS3cw==","expect":"100-continue","x-amz-date":"20240415T234432Z","x-amz-content-sha256":"UNSIGNED-PAYLOAD","authorization":"AWS4-HMAC-SHA256 ...","content-length":"5389"} Error: Not a directory`

--- a/src/sdk/namespace_fs.js
+++ b/src/sdk/namespace_fs.js
@@ -736,7 +736,7 @@ class NamespaceFS {
                         cached_dir = await dir_cache.get_with_cache({ dir_path, fs_context });
                     }
                 } catch (err) {
-                    if (err.code === 'ENOENT') {
+                    if (['ENOENT', 'ENOTDIR'].includes(err.code)) {
                         dbg.log0('NamespaceFS: no keys for non existing dir', dir_path);
                         return;
                     }
@@ -896,6 +896,7 @@ class NamespaceFS {
             await this._check_path_in_bucket_boundaries(fs_context, file_path);
             await this._load_bucket(params, fs_context);
             let stat = await nb_native().fs.stat(fs_context, file_path);
+
             const isDir = native_fs_utils.isDirectory(stat);
             if (isDir) {
                 if (!stat.xattr?.[XATTR_DIR_CONTENT]) {
@@ -2353,7 +2354,7 @@ class NamespaceFS {
 
     _translate_object_error_codes(err) {
         if (err.rpc_code) return err;
-        if (err.code === 'ENOENT') err.rpc_code = 'NO_SUCH_OBJECT';
+        if (err.code === 'ENOENT' || err.code === 'ENOTDIR') err.rpc_code = 'NO_SUCH_OBJECT';
         if (err.code === 'EEXIST') err.rpc_code = 'BUCKET_ALREADY_EXISTS';
         if (err.code === 'EPERM' || err.code === 'EACCES') err.rpc_code = 'UNAUTHORIZED';
         if (err.code === 'IO_STREAM_ITEM_TIMEOUT') err.rpc_code = 'IO_STREAM_ITEM_TIMEOUT';
@@ -2462,6 +2463,10 @@ class NamespaceFS {
             dbg.error('check_access: error ', err.code, err, dir_path, this.bucket_path);
             const is_bucket_dir = dir_path === this.bucket_path;
 
+            if (err.code === 'ENOTDIR' && !is_bucket_dir) {
+                dbg.warn('check_access: the path', dir_path, 'is not a directory');
+                return true;
+            }
             // if dir_path is the bucket path we would like to throw an error
             // for other dirs we will skip
             if (['EPERM', 'EACCES'].includes(err.code) && !is_bucket_dir) {
@@ -2497,6 +2502,10 @@ class NamespaceFS {
                 return false;
             }
         } catch (err) {
+            if (err.code === 'ENOTDIR') {
+                dbg.warn('_is_path_in_bucket_boundaries: the path', entry_path, 'is not a directory');
+                return true;
+            }
             // Error: No such file or directory
             // In the upload use case, the destination file desn't exist yet, need to validate the parent dirs path.
             if (err.code === 'ENOENT') {

--- a/src/test/unit_tests/test_bucketspace.js
+++ b/src/test/unit_tests/test_bucketspace.js
@@ -41,6 +41,7 @@ const is_nc_coretest = process.env.NC_CORETEST === 'true';
 
 const encoded_xattr = 'unconfined_u%3Aobject_r%3Auser_home_t%3As0%00';
 const decoded_xattr = 'unconfined_u:object_r:user_home_t:s0\x00';
+const generic_obj_body = 'AAAABBBBBCCCCCCDDDDD';
 
 // currently will pass only when running locally
 mocha.describe('bucket operations - namespace_fs', function() {
@@ -476,16 +477,8 @@ mocha.describe('bucket operations - namespace_fs', function() {
     });
     mocha.it('list buckets with uid, gid', async function() {
         // Give s3_correct_uid access to the required buckets
-        await Promise.all(
-            [bucket_name]
-            .map(bucket => generate_s3_policy('*', bucket, ['s3:*']))
-            .map(generated =>
-                rpc_client.bucket.put_bucket_policy({
-                    name: generated.params.bucket,
-                    policy: generated.policy,
-                })
-            )
-        );
+        const generated = generate_s3_policy('*', bucket_name, ['s3:*']);
+        await rpc_client.bucket.put_bucket_policy({ name: generated.params.bucket, policy: generated.policy });
         const res = await s3_correct_uid.listBuckets({});
         console.log(inspect(res));
         const list_ok = bucket_in_list([bucket_name], [], res.Buckets);
@@ -511,7 +504,7 @@ mocha.describe('bucket operations - namespace_fs', function() {
 
     mocha.it('put object with out uid gid', async function() {
         try {
-            const res = await s3_owner.putObject({ Bucket: bucket_name + '-s3', Key: 'ob1.txt', Body: 'AAAABBBBBCCCCCCDDDDD' });
+            const res = await s3_owner.putObject({ Bucket: bucket_name + '-s3', Key: 'ob1.txt', Body: generic_obj_body });
             console.log(inspect(res));
             assert.fail('unpreviliged account could put object on nsfs bucket');
         } catch (err) {
@@ -524,9 +517,8 @@ mocha.describe('bucket operations - namespace_fs', function() {
         const s3_client_wrong_uid = generate_s3_client(access_key.unwrap(), secret_key.unwrap());
         const bucket = bucket_name + '-s3';
         const key = 'ob1.txt';
-        const body = 'AAAABBBBBCCCCCCDDDDD';
         try {
-            const res = await s3_client_wrong_uid.putObject({ Bucket: bucket, Key: key, Body: body });
+            const res = await s3_client_wrong_uid.putObject({ Bucket: bucket, Key: key, Body: generic_obj_body });
             console.log(inspect(res));
             assert.fail('unpreviliged account could upload object on nsfs bucket');
         } catch (err) {
@@ -534,17 +526,43 @@ mocha.describe('bucket operations - namespace_fs', function() {
         }
     });
 
+    mocha.it('putObject - regular object - head/get object - key/ - should fail with 404', async function() {
+        const { access_key, secret_key } = account_correct_uid.access_keys[0];
+        const s3_client = generate_s3_client(access_key.unwrap(), secret_key.unwrap());
+        const bucket = bucket_name + '-s3';
+        const key = 'reg_obj.txt';
+        const invalid_dir_key = `${key}/`;
+        const policy = generate_s3_policy('*', bucket, ['s3:*']);
+        await rpc_client.bucket.put_bucket_policy({ name: bucket, policy: policy.policy });
+        await s3_client.putObject({ Bucket: bucket, Key: key, Body: generic_obj_body });
+        try {
+            await s3_client.headObject({ Bucket: bucket, Key: invalid_dir_key });
+            assert.fail('head operation successfull although it should have failed');
+        } catch (err) {
+            assert.deepStrictEqual(err.name, 'NotFound');
+        }
+        try {
+            await s3_client.getObject({ Bucket: bucket, Key: invalid_dir_key });
+            assert.fail('get operation successfull although it should have failed');
+        } catch (err) {
+            assert.deepStrictEqual(err.Code, 'NoSuchKey');
+        }
+        const list_res = await s3_client.listObjects({ Bucket: bucket, Prefix: invalid_dir_key, Delimiter: '/' });
+        assert.equal(object_in_list(list_res, key), false);
+        assert.equal(object_in_list(list_res, invalid_dir_key), false);
+        await s3_client.deleteObject({ Bucket: bucket, Key: key });
+    });
+
     mocha.it('putObject/head object - xattr invalid URI', async function() {
         const { access_key, secret_key } = account_correct_uid.access_keys[0];
         const s3_client = generate_s3_client(access_key.unwrap(), secret_key.unwrap());
         const bucket = bucket_name + '-s3';
         const key = 'obj_dot_xattr.txt';
-        const body = 'AAAABBBBBCCCCCCDDDDD';
         const xattr = { 'key1.2.3': encoded_xattr };
         const policy = generate_s3_policy('*', bucket, ['s3:*']);
         await rpc_client.bucket.put_bucket_policy({ name: bucket, policy: policy.policy });
 
-        await s3_client.putObject({ Bucket: bucket, Key: key, Body: body, Metadata: xattr });
+        await s3_client.putObject({ Bucket: bucket, Key: key, Body: generic_obj_body, Metadata: xattr });
         const head_res = await s3_client.headObject({ Bucket: bucket, Key: key });
 
         assert.deepStrictEqual(head_res.Metadata, xattr);
@@ -607,7 +625,7 @@ mocha.describe('bucket operations - namespace_fs', function() {
             Key: 'ob1.txt',
             UploadId: res1.UploadId,
             PartNumber: 1,
-            Body: 'AAAABBBBBCCCCCCDDDDD',
+            Body: generic_obj_body,
         });
         try {
             // list_multiparts
@@ -623,7 +641,7 @@ mocha.describe('bucket operations - namespace_fs', function() {
         const bucket = bucket_name + '-s3';
         const policy = generate_s3_policy('*', bucket, ['s3:*']);
         await rpc_client.bucket.put_bucket_policy({ name: bucket, policy: policy.policy });
-        const res = await s3_correct_uid.putObject({ Bucket: bucket_name + '-s3', Key: 'ob1.txt', Body: 'AAAABBBBBCCCCCCDDDDD' });
+        const res = await s3_correct_uid.putObject({ Bucket: bucket_name + '-s3', Key: 'ob1.txt', Body: generic_obj_body });
         assert.ok(res && res.ETag);
     });
 
@@ -653,7 +671,7 @@ mocha.describe('bucket operations - namespace_fs', function() {
         config.NSFS_CALCULATE_MD5 = true;
         const key = 'ob11.txt';
         const source_key = key;
-        const put_res = await s3_correct_uid.putObject({ Bucket: bucket_name + '-s3', Key: source_key, Body: 'AAAABBBBBCCCCCCDDDDD' });
+        const put_res = await s3_correct_uid.putObject({ Bucket: bucket_name + '-s3', Key: source_key, Body: generic_obj_body });
         assert.ok(put_res && put_res.ETag);
         await s3_correct_uid.copyObject({ Bucket: bucket_name + '-s3', Key: key, CopySource: bucket_name + `-s3/${source_key}` });
         config.NSFS_CALCULATE_MD5 = false;
@@ -933,7 +951,7 @@ mocha.describe('list objects - namespace_fs', async function() {
     const s3_b_name = 's3-created-bucket';
     const s3_root_b_name = 's3-root-created-bucket';
 
-    const body = 'AAAABBBBBCCCCCCDDDDD';
+    const body = generic_obj_body;
     const accounts = {
         'account1': { uid: 2000, gid: 2000 },
         'account2': { uid: 2001, gid: 2001 },

